files/image_config/platform/rc.local

```diff
@@ -85,6 +88,51 @@ update_mgmt_interface_macaddr() {
     sed -i "/eth0/ s/ATTR{address}==\"$old_mac\"/ATTR{address}==\"$new_mac\"/g" /etc/udev/rules.d/70-persistent-net.rules
 }
 
+firsttime_exit() {
+    rm -rf $FIRST_BOOT_FILE
+    exit 0
+}
+
+# Given a string of tuples of the form field=value, extract the value for a field
+# In : $string, $field
+# Out: $value
+value_extract() {
+    set -- $string
+    for x in "$@"; do
+        case "$x" in
+            $field=*)
+                value="${x#$field=}"
+        esac
+    done
+}
+
+# Set up previous and next reboot cause files
+process_reboot_cause() {
+    REBOOT_CAUSE_FILE="/var/cache/sonic/reboot-cause.txt"
+    PREVIOUS_REBOOT_CAUSE_FILE="/var/cache/sonic/previous-reboot-cause.txt"
+
+    # Set the previous reboot cause accordingly
+    # If this is the first boot after an image upgrade, state that as the
+    # cause. Otherwise, move REBOOT_CAUSE_FILE to PREVIOUS_REBOOT_CAUSE_FILE.
+    # REBOOT_CAUSE_FILE should always exist, but we add the else case
+    # to ensure we always generate PREVIOUS_REBOOT_CAUSE_FILE here
+    if [ -f $FIRST_BOOT_FILE ]; then
+        echo "SONiC firmware upgrade" > $PREVIOUS_REBOOT_CAUSE_FILE
```

Comment:
>SONiC firmware upgrade [](start = 14, length = 22)

How about "First boot". "Firmware: is confusing.

Reply:
I think "first boot" is far more confusing to an end-user, especially if the device has been in production for a while and was recently upgraded. If you don't like "firmware," I would suggest `SONiC operating system upgrade`, `SONiC OS upgrade` or `SONiC image upgrade` as alternatives.

Reply:
I see. Then 'upgrade' is also confusing. Technically the first SONiC installation is not a upgrade.

Maybe the general "Unknown reboot cause" is good enough, since the currently implementation know nothing  about first reboot and sonic upgrade.

---
In reply to: [197971849](https://github.com/Azure/sonic-buildimage/pull/1812#discussion_r197971849) [](ancestors = 197971849)

Reply:
I see your point, but I think it's more useful to know that the last time the device rebooted was due to a firmware upgrade. Maybe `SONiC image installation`?